File: src/data/courses/dsa/trees/structural-property-problems/check-if-balanced/code/python/check_if_balanced.py
# ============================================================
# Balanced Binary Tree
# ============================================================

class Solution1:
    def isBalanced(self, root) -> bool:
        if not root: return True
        lh = self.height(root.left)
        rh = self.height(root.right)
        return abs(lh-rh)<=1 and self.isBalanced(root.left) and self.isBalanced(root.right)
    def height(self, node):
        if not node: return 0
        return 1 + max(self.height(node.left), self.height(node.right))

class Solution2:
    def isBalanced(self, root) -> bool:
        return self.check(root) != -1
    def check(self, node):
        if not node: return 0
        l = self.check(node.left)
        if l == -1: return -1
        r = self.check(node.right)
        if r == -1: return -1
        if abs(l - r) > 1: return -1
        return 1 + max(l, r)

File: src/data/courses/dsa/trees/structural-property-problems/check-if-balanced/code/java/CheckIfBalanced.java
// ============================================================
// Balanced Binary Tree
// ============================================================

class Solution1 {
    public boolean isBalanced(TreeNode root) {
        if (root == null) return true;
        return Math.abs(height(root.left)-height(root.right))<=1 && isBalanced(root.left) && isBalanced(root.right);
    }
    int height(TreeNode n) { return n==null ? 0 : 1+Math.max(height(n.left),height(n.right)); }
}

class Solution2 {
    public boolean isBalanced(TreeNode root) { return check(root) != -1; }
    int check(TreeNode n) {
        if (n==null) return 0;
        int l = check(n.left); if (l==-1) return -1;
        int r = check(n.right); if (r==-1) return -1;
        if (Math.abs(l-r)>1) return -1;
        return 1+Math.max(l,r);
    }
}
